Add issue/PR templates

BytewaveMLP opened the pull request:
Currently just using similar templates as added to PonyGames in Tschrock/PonyGames#10. Needs to be reviewed; we probably want to collect more information and be more noob-friendly. Also may be useful to have separate PR templates for gameplay and internals changes.

Closes #114.

Tschrock commented on .github/ISSUE_TEMPLATE/feature.md:
```diff
@@ -0,0 +1,21 @@
+---
+name: Feature request
+about: Suggest new features to be included in pixel.horse.
+labels: enhancement
```
This should be "feature"

mathieucaroff commented:
I've just used both the pull request template and the issue template today.

I'm in favor of providing (open) guideline to inform (and possibly reassure) the users and developers posting issues. However, I'm not favorable to forcing them into a predefined (closed) template. Fulfilling some of the tasks required by a template can be hard depending on the reporter situation. Issues can take various forms so it's not possible to guarantee that any specific information will always be useful. This is why all parts of a template should be optional.

I agree with suggesting the reporters to provide some specific information, but they should not feel obliged to do so. I think they will be more intelligent when faced with an open guideline than with a closed template. If they miss any important information, these will be obtained through discussing the issue with them.

Communication is important.

Now this is just my feeling - maybe I'm the only one who has a problem with heavy templates -- they make me feel like I'm doing something wrong when I discard their demands, while I know it is fine to do so.

@BytewaveMLP @Tschrock How do you feel about issues templates? What do you think of my perception of things?

BytewaveMLP commented:
I'm not sure I agree. Perhaps an "additional comments, questions, concerns" field should be provided, but I think these forms provide a structured way for people to provide pertinent information in diagnosing and fixing issues. Users are always welcome to provide additional context and information, but I'd say the majority of reports would fit these templates.

Tschrock commented:
@mathieucaroff My thoughts on templates:

For bugs, I like the semi-structured template that most repos seem to have (and that we've replicated here):
 - check for dupes
 - describe issue
 - give reproduction steps and expected behavior
 - give environment information

I also like the idea of having a few other notes encouraging people to include screenshots, error logs, etc (only when applicable)

The one part that can get weird sometimes is the repro steps. Depending on the issue, they could be nauseatingly mundane (What's the repro steps for "Clicking on this button doesn't work"? ..."Try clicking on the button"). They could also be effectively indescribable (How do you reproduce "I can't access the admin panel"? "Uhhhh, set up a server exactly like mine?") - though I feel like those are more likely to be "support tickets" than bug reports.

For feature requests... well I don't think I've ever seen a feature request template I really liked. They usually make some assumptions about the type of request or how the requester is thinking about it. For example: the one here feels purely focused on "what problem does this solve?" - but that can kind of give the wrong impression (it feels like you need to have a definitive, "real" problem that this request would "fix", as opposed to just wanting to improve the project or suggest a cool idea).

For pull requests, I like the simple format that's done here. There's just 3 basic questions - Is there a related issue? What have you changed? If it's a visual change, do you have any pictures of what you changed?

TLDR:
:heavy_check_mark: Bug template
:heavy_multiplication_x: Feature template
:heavy_check_mark: PR template

BytewaveMLP commented:
For the feature request template, should we remove it entirely? Reword it to sound less like it's focused on problem-solving? I'm a little concerned that someone'll hit "new issue," only see "bug report," and think feature requests don't belong here.

As for the bug report template, @mathieucaroff I've addressed your comment, but I realize there's a lot of changes that we push that don't increment the version number. I'd preferably like them to provide a commit sha1, but that's probably too much to get out of some.

blackhole12 commented:
I disagree on the feature request comments here. A feature request template is still useful, but it shouldn't be focusing on what any individual feature "should" have. Instead, it should be asking what _kind_ of feature it is. Is it related to new cosmetics? Is it gameplay? Is it a new tool? Is it an improvement to an existing tool? Is it a codebase refactor?

Feature requests should have the same "check for a duplicate" line that bugs do, because duplicate feature requests are extremely common.

mathieucaroff commented:
@BytewaveMLP (commit sha1). Asking costs nothing. Requiring may cost missing bugs.

Let's ask reporters to "provide if possible" - let's not force them to.

mathieucaroff commented:
About Issue templates. I'm not against them as a whole. They provide guidelines for issue reporters, which is good. However, I'm against making following the template mandatory (and thus, constraining to the reporters). The reporters are already making an effort to let us know about something we don't. Let's accept their effort politely and not reject it because it's not "good enough".

I'm in favor of having recommendations, not demands.

mathieucaroff commented on .github/ISSUE_TEMPLATE/bug.md:
```diff
@@ -0,0 +1,35 @@
+<!-- Place an x within the brackets to denote you've completed the checklist item -->
+
+  - [ ] I have verified this is an issue with the site itself, and not just my system
+  - [ ] I have searched for duplicate issues
```
How do I know that I've searched hard enough ?

How do I know whether or not "issue Y" is a duplicate of mine? What to do if I'm not sure?

Tschrock replied:
How do you know that you've done anything well enough? You don't really. You just try your best

But to actually answer your question, searching through the existing list of bugs is sufficient. Plus GitHub will show similar issues as you're creating one, so you have an extra chance to see a potential dupe. How do you know if it's a dupe? That's a question that doesn't have a definitive answer unless you know what's all going on. What to do if I'm not sure? Create a new one, or leave a comment on the existing one

This is a very simple question and I think you're thinking about it too hard

BytewaveMLP replied:
Gonna have to agree with @Tschrock. It's difficult for us to quantify exactly how much effort should be put in. The reporter should just put their Best Effort™ into searching for potential duplicates—but, hell, even GitHub's duplicate issue suggestion is better than nothing.

This checkbox is just to make a user stop and think about it for a moment before reporting. If they spot a related issue in the GitHub suggestions, the hope is they'll move to +1 that instead of opening a duplicate.

mathieucaroff commented on .github/ISSUE_TEMPLATE/bug.md:
```diff
@@ -0,0 +1,35 @@
+
+  - [ ] I have verified this is an issue with the site itself, and not just my system
+  - [ ] I have searched for duplicate issues
+  - [ ] (optional) I will link to a PR with a proposed fix in the issue comments
```
---

Why the hay use imperative style for an optional matter ? This is confusing. I'd say it needs to be move out of paragraph and rephrased.

mathieucaroff replied:
Solved in BytewaveMLP's commit 1ad0f58e29aeeb741a45da5635ae9d8ec95c5ec2.

mathieucaroff commented on .github/ISSUE_TEMPLATE/bug.md:
```diff
@@ -0,0 +1,35 @@
+## Steps to reproduce
+
+  1. [Step one]
+  2. [Step two]
```
---

Let's talk more about *why* we need to be able to reproduce, and just suggest to list steps to reproduce. That way, the reporter can provide useful instructions, if necessary.

If you think that they may not know about markdown features, then providing a template for this seems like the right solution, but keep it optional.

Tschrock replied:
Grammar: this should say "*list* the steps to reproduce"

Perhaps the title could be reworded or the section split apart between the root questions of "what causes this?" and "what do you need to do to trigger/reproduce this?"

mathieucaroff replied:
&grammar - thank you

mathieucaroff commented on .github/ISSUE_TEMPLATE/bug.md:
```diff
@@ -0,0 +1,35 @@
+
+**Expected behavior:** [What should have happened?]
+
+**Actual behavior:** [What actually happened?]
```
---

Expected behavior and Actual behavior work together. We should just ask about the differences between what they expected and what they got. This is more flexible for them to present it in a meaningful way.

Tschrock replied:
Isn't asking about what you expect and what you got and asking what the difference is the same thing? You're going to have to explain yourself in a bit more detail here :/

mathieucaroff replied:
The difference between the two questions and my single question is that **formally**, a single question is more flexible, and gives the reporter more freedom to present its problem. Now if both cases are similar in substance, it means the change should not pose any problem ; )

~Here's a very poor example of what I mean:~
~What is `a`? It is `{ color: 'white', kind: 'curved', size: 8 }`~
~What is `b`? It is `{ color: 'white', kind: 'curved', size: 6 }`~
~What is the difference (delta) between `a` and `b`? It is `{ size: 2 }`~

mathieucaroff replied:
Here is my proposition for this part of the template:

```md
<!-- Please describe in context the difference between the current behavior and the expected behavior -->
```

mathieucaroff commented on .github/ISSUE_TEMPLATE/bug.md:
```diff
@@ -0,0 +1,35 @@
+
+**pixel.horse version:** [Can be found in the bottom left of the home page, or on the about page]
+
+**Node.js version:** [(Only needed if you're running your own instance of the site)]
```
> Browser/OS version
> pixel.horse version
> Node.js version

Why do we need these? (Let's explain -)

---

Depending on who you ask, you might find that asking about which OS some one uses touches to private life -- more than it touches to bug reporting, which can participate to deter some reporters.

Tschrock replied:
OS/Version is a standard bug report question, and I've never heard of it "touching to private life". If you have some examples of why you're thinking this way please let me know

mathieucaroff replied:
Well nothing new here, but OSes are linked with tech giants (or not), have a price, restrict which software can be run (or can be run easily) on a the device, what activities the user will have with their device, and how they are going to spend their time using it. These information contribute (with various intensity depending on who you consider) to the judgement people make of each other. This is what leads me to say some users might consider it private.
My conclusion is that we should not require it if we don't **need** it. I'm OK with recommending the reporter to provide it.

Tschrock replied:
>restrict which software can be run

This reminds me, OS will also be very helpful in knowing what type of tools and methods to suggest when trying to help someone with an issue, especially hosting-related ones

mathieucaroff replied:
Definitely, yes

BytewaveMLP replied:
I see no reason why we *wouldn't* need the OS the reporter is running. A number of issues can be traced back to a specific browser/OS configuration, and if we can't reproduce the issue ourselves, knowing what specific system configuration the issue occurred on would save us from asking additional questions.

The OS a system is running isn't private information. Most every browser submits it as part of the UA string to every site the user visits. If people are judgmental about the OS someone else uses, they're being a dick—and if they express it publicly on an issue page, they're looking to get blocked.

mathieucaroff replied:
Ultimately, this is the question of whether all possibly relevant information(s) shall be provided initially, or if interactions shall be privileged to obtain them.

I'm resolving the conversation, as I'm out of rational arguments.

mathieucaroff commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,13 @@
+## Is this related to an existing issue? Link below:
```
If this is not related to an existing issue, should I just delete the paragraph ?
If it is related to an existing issue, should I just leave the title as is ?

mathieucaroff replied:
See [#discussion_r326601562](#discussion_r326601562), below

mathieucaroff commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,13 @@
+## Describe the changes you've made
+
+  - [Change one]
+  - [Change two]
```
What level of detail is expected?

Tschrock replied:
Basically, things that would show up in a changelog

mathieucaroff replied:
The last Bytewave's comit (1ad0f58e29aeeb741a45da5635ae9d8ec95c5ec2) (2019-09-19) makes it good enough.

mathieucaroff commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,13 @@
+
+## Screenshots of your new feature (if applicable)
+
+[only for gameplay or sprite changes and additions]
```
Should I delete the paragraph if screenshots are not applicable?

mathieucaroff replied:
Could something like we have the following at the beginning or the end of bug reports and feature request
"You are welcome to ignore and remove any part of the issue template which does not apply to your issue, or which poses too much inconvenience to complete. If the removed parts appears necessary or useful, the pony town team will ask out."

If not, say no --

BytewaveMLP replied:
It doesn't really matter if the heading remains or not of it isn't filled out, I'd say. If it's not applicable, it's not applicable. Write `N/A`, remove it, leave it default... if there's no concept work or artwork to submit with the PR, it doesn't affect readability to include/not include this section.

mathieucaroff replied:
I just thought that in my header proposal, "pony town" should be replaced by "pixel horse". Other than that, do you agree with adding the message, or a message of the same substance as header or footer ?

Tschrock commented on .github/ISSUE_TEMPLATE/bug.md:
```diff
@@ -0,0 +1,35 @@
+
+<!-- Place an x within the brackets to denote you've completed the checklist item -->
+
+  - [ ] I have verified this is an issue with the site itself, and not just my system
```
We should enumerate that you can: ask a friend, test on a different browser, or test on a different device
Any of those three should be sufficient for filtering out basic not-our-issue type bugs without causing too much work for the reporter.

Tschrock replied:
Actually, testing another browser may indicate a browser-specific bug, which we would still want to hear about :thinking:

mathieucaroff commented:
@BytewaveMLP the changes made already make the templates a lot better. There remains the checkboxes which could be more specific in how to validate them and how much effort to invest, the rest of the changes are minor and will be sorted out by Tyler Schrock and I, or dismissed.

BytewaveMLP commented:
Merging for now, as this has sat in debate hell for a while. We can always tweak these later, but they're a good foundation.
